src/packages/utils/src/things/json-rpc/json-rpc-quantity.ts

```diff
@@ -1,87 +1,132 @@
-      let val = this.value.slice(2).replace(/^(?:0+(.+?))?$/, "$1");
-      if (val === "" || val === "0") {
-        return BUFFER_EMPTY;
-      }
-    } else if (this.value === 0 || this.value === 0n) {
+    if (this.bufferValue == null) {
       return BUFFER_EMPTY;
     }
 
-    return super.toBuffer();
-  }
-  public toBigInt(): bigint | null {
-    const value = this.value;
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
 
-    // TODO(perf): memoize this stuff
-    if (Buffer.isBuffer(value)) {
-      const bigInt = bufferToBigInt(value);
-      return bigInt == null ? (this._nullable ? null : 0n) : bigInt;
+    if (firstNonZeroByte > 0) {
+      return this.bufferValue.slice(firstNonZeroByte);
     } else {
-      return value == null ? (this._nullable ? null : 0n) : BigInt(value);
+      return this.bufferValue;
+    }
+  }
+
+  public toBigInt(): bigint | null {
+    if (this.bufferValue == null) {
+      return this._nullable ? this.bufferValue : 0n;
+    }
+    if (this.bufferValue.length === 0) {
+      return 0n;
     }
+    return bufferToBigInt(this.bufferValue);
   }
+
   public toNumber() {
-    // TODO(perf): memoize this stuff
-    return typeof this.value === "number"
-      ? this.value
-      : Number(this.toBigInt());
+    if (this.bufferValue == null) {
+      return this._nullable ? this.bufferValue : 0;
+    }
+
+    const firstNonZeroByte = this.findFirstNonZeroByteIndex();
+
+    const length = this.bufferValue.length - firstNonZeroByte;
+    if (length === 0) {
+      return 0;
+    }
+
+    let result: number;
+    if (length > 6) {
+      const trimmedBuffer = firstNonZeroByte === 0 ? this.bufferValue : this.bufferValue.slice(this.bufferValue);
+      result = Number(bufferToBigInt(trimmedBuffer));
+    } else {
+      result = this.bufferValue.readUIntBE(firstNonZeroByte, length);
+    }
+    if (!Number.isSafeInteger(result)) {
+      console.warn(`0x${this.bufferValue.toString("hex")} is too large - the maximum safe integer value is 0${Number.MAX_SAFE_INTEGER.toString(16)}`);
+    }
+    return result;
   }
+
   valueOf(): bigint {
-    const value = this.value;
-    if (value === null) {
-      return value as null;
-    } else if (value === undefined) {
-      return value as undefined;
+    if (this.bufferValue == null) {
+      return <null>this.bufferValue;
     } else {
       return this.toBigInt();
     }
   }
+
+  private findFirstNonZeroByteIndex(): number {
+    let firstNonZeroByte = 0;
+    for (firstNonZeroByte = 0; firstNonZeroByte < this.bufferValue.length; firstNonZeroByte++) {
+      if (this.bufferValue[firstNonZeroByte] !== 0) break;
+    }
+    return firstNonZeroByte;
+  }
+
+  static toBuffer(value: JsonRpcInputArg, nullable?: boolean): Buffer {
+    return Quantity.from(value, nullable).toBuffer();
+  }
+
+  static toString(value: JsonRpcInputArg, nullable?: boolean): string {
+    return Quantity.from(value, nullable).toString();
```

Comment:
I started reading through nodejs Buffer pooling implementation, and I have some thoughts about how we can benefit from bypassing this and providing our own pooling, potentially in all cases, not just the statics.

I'd like to explore this outside of this PR, as it could take some time!